[SEO-95] Add ability to rerun failed tests when rerunning workflow

magni- opened the pull request:
This was tested in https://github.com/rainforestapp/github-action/actions/runs/1531060148

magni- committed:
Remove accidental reference to crowd parameter type

This was copy-pasted in from the Orb repo, where parameters have types. There is no such concept for GitHub actions (all parameters are strings), so I removed the type definitions from the README--except this one which I missed, until now.

magni- committed:
Add support for rerun flow via cache fork

The CircleCI Orb uses CircleCI caching for this. This won't work with the standard cache action for a couple reasons:
1. actions/cache only writes to the cache if there was no cache hit (so rereruns won't work)
2. actions/cache only writes to the cache if the workflow passed (which defeats the whole point, since we want to write when runs _fail_).

Instead, we use a fork that allows us to always cache.

magni- committed:
Version 1.0.0!

magni- commented:
/reviewme @jbarber @rainforestapp/devs

magni- committed:
Update README

magni- commented on action.yml:
```diff
@@ -144,6 +159,13 @@ runs:
+      if: failure()
+      shell: bash
+      run: |
+        sudo apt-get install libxml2-utils
```
I'll update this PR to use the new `--save-run-id` flag once that's shipped https://github.com/rainforestapp/rainforest-cli/pull/442 so that we no longer need to install `libxml2-utils` here.

jbarber commented on action.yml:
```diff
@@ -74,7 +81,15 @@ runs:
+
+          RUN_COMMAND="rerun ${RAINFOREST_RUN_ID} --skip-update --token ${{ inputs.token }} --junit-file results/rainforest/junit.xml"
+        else
+          RUN_COMMAND="run --skip-update --token ${{ inputs.token }} --run-group ${{ inputs.run_group_id }} --junit-file results/rainforest/junit.xml"
```
```suggestion
          RUN_COMMAND="run --skip-update --token \"${{ inputs.token }}\" --run-group \"${{ inputs.run_group_id }}\" --junit-file results/rainforest/junit.xml"
```

jbarber commented on action.yml:
```diff
@@ -74,7 +81,15 @@ runs:
+          RAINFOREST_RUN_ID=$(cat .rainforest_run_id)
+          echo "Rerunning Run ${RAINFOREST_RUN_ID}"
+
+          RUN_COMMAND="rerun ${RAINFOREST_RUN_ID} --skip-update --token ${{ inputs.token }} --junit-file results/rainforest/junit.xml"
```
```suggestion
          RUN_COMMAND="rerun \"${RAINFOREST_RUN_ID}\" --skip-update --token \"${{ inputs.token }}\" --junit-file results/rainforest/junit.xml"
```

jbarber commented on README.md:
```diff
@@ -98,6 +96,9 @@ Set to `true` to run parameter validations without actually starting a run in Ra
 If no `dry_run` parameter is passed in, the run will be started in Rainforest.
 
+## Rerunning failed tests
+The Rainforest QA GitHub Action uses GitHub [caching](https://docs.github.com/en/actions/advanced-guides/caching-dependencies-to-speed-up-workflows) to know when a workflow is [being rerun](https://docs.github.com/en/actions/managing-workflow-runs/re-running-workflows-and-jobs). It will automatically rerun only the tests which failed in the previous run.
```
I think this could do with a bit more context about why to do this and how we expect users to use it - we should give an opinion about how they use the action in their workflows. Not sure if this context belongs in this PR or in this part of the README.

magni- replied:
Good point 👍🏼 For the orb, that context was hidden away in the [blog post](https://rainforest.engineering/2021-01-20-shipping-faster-orb/) detailing how to use the feature. I'll add a little bit of copy here to make it more obvious you shouldn't rerun failed workflows without checking results/fixing tests first.

jbarber reviewed (approved):
Please consider if we can give users more of an opinion about when to rerun and the workflow, probably something like:
* See there's a failure
* Check if it's a real failure or not
* Rerun if it's a real failure

Also check the quoting of arguments for shell commands.

magni- commented:
> Also check the quoting of arguments for shell commands.

Why are those quotes necessary? Both the token's and the run ID's presence are validated before they're added to the command string.

jbarber commented:
> > Also check the quoting of arguments for shell commands.
> 
> Why are those quotes necessary? Both the token's and the run ID's presence are validated before they're added to the command string.

I'm not familiar with the permissions for actions and who is running what, so wanted to flag them as a concern. For example, can someone in the GH org set `inputs.run_group_id` so that it causes a shell escape and displays the `inputs.token`?

`RAINFOREST_RUN_ID` is set by reading a file, and we don't validate the format of the contents of the file, so would it be possible for an attacker to write a string to `.rainforest_run_id` that would cause badness?

etc. etc.

jbarber committed:
Add defensive quoting

https://github.com/rainforestapp/github-action/pull/3#issuecomment-988005849

magni- committed:
Fix base case test

This was passing because bash's -e doesn't work for errors in conditionals...
https://unix.stackexchange.com/questions/407716/set-e-does-not-terminate-script-when-error-occurs-in-conditional
